File: graphs/reducers/prioritize.js
// Deterministic prioritisation of audit issues: severity weight, then category weight; keep the top k.
// input: { items: [{severity, category, ...}] }  args: { k: 12, weights?: {blocker: 100, major: 30, minor: 8, nit: 1} }
export default async function prioritize(input, args) {
  const items = Array.isArray(input.items) ? input.items : [];
  const w = { blocker: 100, major: 30, minor: 8, nit: 1, ...(args.weights ?? {}) };
  const cat = { security: 5, correctness: 4, reliability: 3, performance: 2, maintainability: 1, tests: 1, docs: 0 };
  const scored = items.map((it) => ({ ...it, priority: (w[String(it.severity).toLowerCase()] ?? 1) + (cat[String(it.category).toLowerCase()] ?? 0) }));
  scored.sort((a, b) => b.priority - a.priority);
  const k = Number(args.k ?? 12);
  const kept = scored.slice(0, k).map((it, i) => ({ rank: i + 1, ...it }));
  return { items: kept, dropped: Math.max(0, scored.length - k), by_severity: countBy(scored, "severity"), _stats: { in: items.length, out: kept.length } };
}

function countBy(list, key) {
  const out = {};
  for (const it of list) out[String(it[key] ?? "unknown")] = (out[String(it[key] ?? "unknown")] ?? 0) + 1;
  return out;
}

File: gren/kit/graphs/reducers/prioritize.py
# Deterministic prioritisation of audit issues: severity weight, then category weight; keep the top k.
# input: { items: [{severity, category, ...}] }  args: { k: 12, weights?: {blocker: 100, major: 30, minor: 8, nit: 1} }
from collections import Counter

CATEGORY = {"security": 5, "correctness": 4, "reliability": 3, "performance": 2, "maintainability": 1, "tests": 1, "docs": 0}


def reduce(input, args, ctx):
    items = input.get("items") if isinstance(input.get("items"), list) else []
    w = {"blocker": 100, "major": 30, "minor": 8, "nit": 1, **(args.get("weights") or {})}
    scored = [{**it, "priority": w.get(str(it.get("severity")).lower(), 1) + CATEGORY.get(str(it.get("category")).lower(), 0)} for it in items if isinstance(it, dict)]
    scored.sort(key=lambda it: -it["priority"])
    k = int(args.get("k", 12))
    kept = [{"rank": i + 1, **it} for i, it in enumerate(scored[:k])]
    by_sev = Counter(str(it.get("severity") or "unknown") for it in scored)
    return {"items": kept, "dropped": max(0, len(scored) - k), "by_severity": dict(by_sev), "_stats": {"in": len(items), "out": len(kept)}}
